libbeat/common/file/interval_rotator.go

```diff
@@ -93,7 +96,41 @@ func (r *intervalRotator) initialize() error {
 			return lastInterval != currentInterval
 		}
 	}
-	return nil
+
+	// if rotation is not allowed on startup, find the newest rotated file last modification time
+	// or that of the unrotated log file
+	if !rotateOnStartup {
+		logsDir := filepath.Dir(filename)
+		logfiles, err := ioutil.ReadDir(logsDir)
+		if err != nil && log != nil {
+			log.Debugw("Not attempting to find last rotated time, configured logs dir cannot be opened: %v", err)
+			return
+		}
+
+		r.lastRotate = time.Time{}
+		if len(logfiles) == 1 && logfiles[0].Name() == filename {
+			if log != nil {
+				log.Debugw("Setting last rotated time to the last modification time of the log")
+			}
+
+			r.lastRotate = logfiles[0].ModTime()
+			return
+		}
+		basenamePrefix := filepath.Base(filename) + "-"
+		for _, fi := range logfiles {
+			if strings.HasPrefix(fi.Name(), basenamePrefix) {
+				if fi.ModTime().After(r.lastRotate) {
+					r.lastRotate = fi.ModTime()
+				}
+			}
+		}
+
+		if log != nil {
+			log.Debugw("Set last rotated time to", r.lastRotate)
+		}
+	}
```

Comment:
It would be nice if the part of this code that looks at `logFiles` and computes `r.lastRotate` was extracted into its own function so it could be unit tested.

Reply:
AFAIK unit test should not touch the file system, and I cannot test this algorithm without creating files.

Reply:
I could mock the `ReadDir` function but that seems a bit too much for such.

Reply:
Disregard what I have just written. I am testing it.

Reply:
I don't think you need to mock `ReadDir`. I was thinking extracting the code from https://github.com/elastic/beats/pull/17613/files#diff-dcee89f48c6b9af2141c1852916ff7b0R109 onwards. Essentially, you'd pass your helper function a `[]os.FileInfo`. Since [`os.FileInfo`](https://godoc.org/os#FileInfo) is an interface, you could mock out an implementation for the unit test?

Reply:
Yes, I am not mocking it. You will see what I have done soon. :D

Reply:
Tests are always run in the current directory. Why not just create a temporary directory to run tests in? I think we do this in some other places as well.

There exists also some vfs packges for go. Also overkill for here, at some point we should evaluate some of the libs to see if one matches our needs.